Managers/UInputMan.cpp

```diff
@@ -939,172 +981,180 @@ namespace RTE {
+
+		for (; controllerIndex < s_PrevJoystickStates.size(); ++controllerIndex) {
+			if (s_PrevJoystickStates[controllerIndex].m_DeviceIndex == deviceIndex || s_PrevJoystickStates[controllerIndex].m_DeviceIndex == -1) {
+				if (SDL_IsGameController(deviceIndex)) {
+					SDL_GameController *gameController = SDL_GameControllerOpen(deviceIndex);
+					if (!gameController) {
+						std::string connectString = s_PrevJoystickStates[controllerIndex].m_DeviceIndex == deviceIndex ? "reconnect" : "connect";
+						g_ConsoleMan.PrintString("ERROR: Failed to " + connectString + " Gamepad " + std::to_string(controllerIndex + 1));
+						break;
+					}
+					controller = SDL_GameControllerGetJoystick(gameController);
+				} else {
+					controller = SDL_JoystickOpen(deviceIndex);
+				}
+				if (!controller) {
+					std::string connectString = s_PrevJoystickStates[controllerIndex].m_DeviceIndex == deviceIndex ? "reconnect" : "connect";
+					g_ConsoleMan.PrintString("ERROR: Failed to " + connectString + " Gamepad " + std::to_string(controllerIndex + 1));
+					break;
+				}
+				std::string connectString = s_PrevJoystickStates[controllerIndex].m_DeviceIndex == deviceIndex ? " reconnected" : " connected";
+				g_ConsoleMan.PrintString("INFO: Gamepad " + std::to_string(controllerIndex + 1) + connectString);
+				break;
 			}
 		}
-	}
-
-/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
-
-#ifdef __unix__
-	void UInputMan::HandleAllegroMouseInput() {
-		if (_xwin.display == 0) {
-			return;
-		}
-
-		std::vector<XEvent> events;
-		events.reserve(XPending(_xwin.display) + 10);
-		while(XPending(_xwin.display) > 0) {
-			XEvent e;
-			XNextEvent(_xwin.display, &e);
-			events.push_back(e);
-		}
 
-		m_AllegroMousePreviousX = mouse_x;
-		m_AllegroMousePreviousY = mouse_y;
-		int mouseDeltaX = 0;
-		int mouseDeltaY = 0;
-
-		int halfResX = _xwin.window_width / 2;
-		int halfResY = _xwin.window_height / 2;
-		for (std::vector<XEvent>::reverse_iterator event = events.rbegin(); event < events.rend(); ++event) {
-			switch (event->type) {
-				case MotionNotify: {
-					mouseDeltaX = event->xmotion.x - m_AllegroMousePreviousX;
-					mouseDeltaY = event->xmotion.y - m_AllegroMousePreviousY;
-					_xwin_mouse_interrupt(mouseDeltaX, mouseDeltaY, 0, 0, mouse_b);
-					_mouse_x = m_AllegroMousePreviousX = !m_TrapMousePos ? event->xmotion.x : halfResX;
-					_mouse_y = m_AllegroMousePreviousY = !m_TrapMousePos ? event->xmotion.y : halfResY;
-
-					if (m_TrapMousePos && (mouseDeltaX != 0 || mouseDeltaY != 0)) {
-						XWarpPointer(_xwin.display, _xwin.window, _xwin.window, 0, 0, 0, 0, halfResX, halfResY);
+		if (!controller && controllerIndex == s_PrevJoystickStates.size()) {
+			controllerIndex = 0;
+			for (; controllerIndex < s_PrevJoystickStates.size(); ++controllerIndex) {
+				if (s_PrevJoystickStates[controllerIndex].m_JoystickID == -1) {
+					if (SDL_IsGameController(deviceIndex)) {
```

Comment:
What's up with this code that's basically duplicate of the code above? Can it not be merged into the above stuff?

Also, even if it can't, the meat of these should probably be extracted into a lambda function or private method (even more so before I merged the 2 above). The console printing can be done based on a flag or passed in by the caller or whatever.